File: Day076_2025/Q076_GFG.py
from typing import List

class Solution:
    def isSubsetSum(self, arr: List[int], sum: int) -> bool:
        n = len(arr)
        prev = [False] * (sum + 1)
        curr = [False] * (sum + 1)

        prev[0] = True  # Base case: sum 0 is always possible

        for i in range(1, n + 1):
            for j in range(sum + 1):
                if j < arr[i - 1]:
                    curr[j] = prev[j]
                else:
                    curr[j] = prev[j] or prev[j - arr[i - 1]]
            prev = curr[:]  # Copy current row to previous

        return prev[sum]

File: Day076_2025/Q076_GFG.java
class Q076_GFG {
    public static boolean isSubsetSum(int[] arr, int sum) {
        int n = arr.length;
        boolean[] prev = new boolean[sum + 1];
        boolean[] curr = new boolean[sum + 1];

        prev[0] = true; // Base case: sum 0 is always possible

        for (int i = 1; i <= n; i++) {
            for (int j = 0; j <= sum; j++) {
                if (j < arr[i - 1]) {
                    curr[j] = prev[j];
                } else {
                    curr[j] = prev[j] || prev[j - arr[i - 1]];
                }
            }
            prev = curr.clone(); // Copy current row to previous
        }

        return prev[sum];
    }
}
